### PR TITLE
Serialize expression compilation requests

### LINKED ISSUE #18595: VM evaluateInFrame does not always provide a response
While debugging some expression evaluation issues, I noticed that the Watch window in VS Code sometimes stops responding correctly. Turns out it was because we never provided a response to a request for an evaluation.

I captured a log, and it seems I can reliably repro the VM not responding to a request for an evaluation. It seems to happen when I send multiple in quick succession that are invalid.

Both invalid requests are written to stderr, but only one gives a response via the WebSocket/JSON API.

Given how badly VS Code seems to handle this, I've put a 1sec timeout in for the next version of Dart Code), though this seems like a fairly serious issue.

To repro, I'm using Stocks, I have a breakpoint in the build method of StockRow and have several invalid watch expressions in the watch window.

```
[11:16:19] [Observatory] ==> {"id":"531","method":"evaluateInFrame","params":{"expression":"aaa","frameIndex":0,"isolateId":"isolates/341014419"}}
[11:16:19] [Observatory] ==> {"id":"532","method":"evaluateInFrame","params":{"expression":"bbb","frameIndex":0,"isolateId":"isolates/341014419"}}
[11:16:19] [FlutterRun] ERR compiler message: org-dartlang-debug:synthetic_debug_expression:1:1: Error: Getter not found: 'aaa'.
[11:16:19] [FlutterRun] ERR compiler message: aaa
compiler message: ^^^
[11:16:19] [FlutterRun] ERR compiler message: org-dartlang-debug:synthetic_debug_expression:1:1: Error: The getter 'aaa' isn't defined for the class '#lib1::StockRow'.
[11:16:19] [FlutterRun] ERR compiler message: Try correcting the name to the name of an existing getter, or defining a getter or field named 'aaa'.
compiler message: aaa
compiler message: ^
[11:16:19] [FlutterRun] ERR compiler message: result c32e6c5b-2d6f-487c-ba97-7981ba29d14a
[11:16:19] [FlutterRun] ERR compiler message: org-dartlang-debug:synthetic_debug_expression:1:1: Error: Getter not found: 'bbb'.
[11:16:19] [FlutterRun] ERR 
compiler message: bbb
compiler message: ^^^
[11:16:19] [FlutterRun] ERR compiler message: org-dartlang-debug:synthetic_debug_expression:1:1: Error: The getter 'bbb' isn't defined for the class '#lib1::StockRow'.
[11:16:19] [FlutterRun] ERR compiler message: Try correcting the name to the name of an existing getter, or defining a getter or field named 'bbb'.
compiler message: bbb
compiler message: ^
[11:16:19] [FlutterRun] ERR compiler message: c32e6c5b-2d6f-487c-ba97-7981ba29d14a build/app.dill.incremental.dill 4
[11:16:19] [Observatory] <== {"jsonrpc":"2.0", "result":{"type":"@Error","_vmType":"LanguageError","kind":"LanguageError","id":"objects\/212","message":"'evaluate:source': error: org-dartlang-debug:synthetic_debug_expression:1:1: Error: The getter 'bbb' isn't defined for the class '#lib1::StockRow'.\nTry correcting the name to the name of an existing getter, or defining a getter or field named 'bbb'.\nbbb\n^"},"id":"532"}
[11:17:03] [FlutterRun] ==> [{"id":"7","method":"app.stop","params":{"appId":"768e5e5e-35b8-4f18-be71-565e320e3261"}}]
[11:17:03] [FlutterRun] <== Application finished.
[11:17:03] [FlutterRun] <== [{"id":"7","result":true}]
[11:17:03] [Observatory] Terminating related process 27193...
[11:17:03] [Observatory] Terminating main process...
[11:17:05] Log file ended
```

### PR DESCRIPTION
**Pull request by aam**
Requests to compile expressions for debugger evaluation can come to flutter while previously submitted requests are still being handled. This PR serializes all compilation requests to the compiler.

This fixes https://github.com/flutter/flutter/issues/18595

### CODE REVIEW

**Comment by aam** [comment]
PTAL, hopefully this will help @DanTup.

**Comment by DanTup** [comment]
Fix seems good to me, thanks! :)

Before:

screen shot 2018-06-21 at 8 20 43 am

After:

screen shot 2018-06-21 at 8 19 41 am

**Comment by DanTup** [comment]
@aam Btw, I presume this issue was specific to how evaluations are routed back to Flutter and didn't affect non-Flutter apps? I noticed the fix was in flutter_tools; but I can't repro any issues when running directly on the desktop VM.

File: packages/flutter_tools/lib/src/compile.dart [hunk 1]
```diff
@@ -186,6 +186,45 @@ class KernelCompiler {
...
+  Future<void> run(ResidentCompiler compiler) async {
+    completer.complete(await compiler._recompile(this));
+  }
+}
+
+class _CompileExpressionRequest implements _CompilationRequest {
+  _CompileExpressionRequest(this.completer, this.expression, this.definitions,
+      this.typeDefinitions, this.libraryUri, this.klass, this.isStatic);
+
+  Completer<CompilerOutput> completer;
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/compile.dart]: This field can be in the base classs?

**Review by a-siva — APPROVED** [review]

File: packages/flutter_tools/lib/src/compile.dart (hunk 1 above)
Comment (aam) [reply packages/flutter_tools/lib/src/compile.dart]: Sure, done. PTAL if you have a chance

**Comment by aam** [comment]
@DanTup wrote
> @aam Btw, I presume this issue was specific to how evaluations are routed back to Flutter and didn't affect non-Flutter apps?

That's right. For standalone Dart VM(non-Flutter apps) there is different(conceptually similar) wrapper. We might have to have similar fix applied to that wrapper(kernel service) too.

**Comment by DanTup** [comment]
Looking at the original issue I'm not certain, but here's a guess (possible @aam can correct me). I think this might be because the way the expression compilation is done is by just sending strings over a stream bare and a response is received:

```
>>> expression
<<< response
>>> expression
<<< response
```

Since these aren't wrapped in anything to indicate which request a response relates to, having multiple requests trying to send requests at the same time might mess things up:

```
>>> expression1
>>> expression2
<<< response // which expression is this the response to?
```

My guess is that serialising these into a queue like this was an easy way to fix the issue. It would probably be better if these requests were more structured though (I think we have another open issue somewhere that newlines in an expression will break?). We use something like JSON-RPC for many other tools (like the Flutter daemon), maybe something similar would work here?


(I'm not sure how this is causing your issue though, but I'm not all that familiar with either parts of this code)

**Comment by andrewkolos** [comment]
Thanks for the response! 

> (I'm not sure how this is causing your issue though, but I'm not all that familiar with either parts of this code)

Sorry for not elaborating here. I think you are already familiar with https://github.com/dart-lang/sdk/issues/54911 (writing to a process risks an uncatchable error that will often have no meaningful stacktrace), which is affecting the tool greatly. In the flutter tool, we've [implemented a workaround to this using runZonedGuarded](https://github.com/flutter/flutter/blob/a647a1df3cd4f283c8699e2bfc6d51b305ac6f88/packages/flutter_tools/lib/src/base/process.dart#L280). However, this implementation is based on `IOSink::flush`, and you can only have one outgoing flush at any given time. Said another way, any call to this workaround should be awaited before making another call so that we don't end up with two concurrent flush calls. `ResidentCompiler` is one place in the tool where we make several writes to a process. However, adding `awaits` there results in a performance regression (see aforementioned https://github.com/flutter/flutter/issues/151255). Thinking about this again, I don't think this issue blocks the tool making progress here. I could instead only call the workaround for the last write; it would just be a bit confusing to readers. Sorry if this was hard to follow—in summary, I don't think `ResidentCompiler` being written to process requests sequentially is blocking anything. It's just the reason is not well documented, and I was hoping I could better understand it and possibly document it myself.

**Comment by DanTup** [comment]
Ah gotcha, thanks for the explanation! (and I agree it would be good to document the reasons for this if they're not clear in the code - though it's probably worth some confirmation from @aam or @a-siva if my assumption above is correct before documenting it as such 😄)
